src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php

```diff
@@ -1386,7 +1386,7 @@ public function getDates()
      */
     public function getDateFormat()
     {
-        return $this->dateFormat ?: $this->getConnection()->getQueryGrammar()->getDateFormat();
+        return $this->dateFormat = $this->dateFormat ?: $this->getConnection()->getQueryGrammar()->getDateFormat();
```

Comment:
This should fix your failing test.

```php
public function getDateFormat()
{
    static $dateFormat = null;
    
    return $this->dateFormat ?: ($dateFormat ??= $this->getConnection()->getQueryGrammar()->getDateFormat());
}
```

If you memoize the `date format` inside the method instead you don't have to touch the class property, changing that property during runtime seems like an unexpected side effect anyway so best to aviod it.

Reply:
that will not work. it will cache it for ALL instances, the same

Reply:
you are very much correct, good catch, i did not take into account that this method is on the base model, I was too focused of a single model.

Reply:
@GrahamCampbell can you clarify what you mean? 

The solution by @stancl would allow ModelA to have it's own dateFormat defined and ModelB can still use default with caching to prevent multiple lookups.

see example, or maybe i misunderstand your comment.

```php
class BaseModel 
{
    protected $dateFormat = null;
    
    public function getDateFormat() 
    {
        static $dateFormat = null;
        
        return $this->dateFormat ?: $dateFormat ??= 'Y-m-d H:i:s';
    }
}

class ModelA extends BaseModel 
{
    protected $dateFormat = 'Ymd His';
}

class ModelB extends BaseModel {}

$instanceA = new ModelA;
$instanceB = new ModelB;

dd(
    $instanceA->getDateFormat(), // "Ymd His"
    $instanceB->getDateFormat(), // "Y-m-d H:i:s"
    $instanceA->getDateFormat(), // "Ymd His"
    $instanceB->getDateFormat(), // "Y-m-d H:i:s"
);
```